### PR TITLE
CSCShowerDigi format and producer (HST-I)

### PR DESCRIPTION
**Pull request by dildick**
#### PR description:

In Run-3 we would like to include a counter in the CSC trigger sensitive to hadronic showers. They can indicate the decay of a long-lived particle in the muon endcap system and have the potential of increasing sensitivity in physics searches by a factor 20. 

This PR introduces `CSCShowerDigi` object and producer in the trigger. They can either be "loose", "nominal", or "tight", depending on how many hits were recorded in a chamber. The `CSCShowerDigi` are produced in addition to the regular CSC trigger primitives and are sent to the EMTF. They are therefore of type `CSCDetIdCSCShowerDigiMuonDigiCollection_simCscTriggerPrimitiveDigis__`. The trigger data has already been upgraded to include additional bits for hadronic showers (see [DN-20-016](https://gitlab.cern.ch/tdr/notes/DN-20-016/blob/master/temp/DN-20-016_temp.pdf)). For the CSC->MPC->EMTF path, 4 `HMT` bits per chamber per BX have been allocated. In the EMTF sector processors we're planning a simple logic block that checks the presence of a shower (e.g. one nominal shower, or two loose showers).  For the EMTF->uGMT 3x2 bits are available per BX per trigger sector. In our current design we would use 1 or 2 bits. That will be done in future PRs.

Additional documentation in [DN-20-033](https://gitlab.cern.ch/tdr/notes/DN-20-033/blob/master/temp/DN-20-033_temp.pdf) and in [this presentation](https://indico.cern.ch/event/1006558/contributions/4227898/attachments/2187396/3696178/HadronicShowerTrigger_20210204.pdf).

#### PR validation:

Tested with H->2s->4b/4d samples  below
https://cmsweb.cern.ch/das/request?view=list&limit=50&instance=prod%2Fglobal&input=%2FHTo2LongLivedTo4b*%2FRun3Winter20DRPremixMiniAOD-110X_mcRun3_2021_realistic_v6-v2%2FGEN-SIM-RAW

Also tested with WF 11634.0.

```
11634.0_TTbar_14TeV+2021+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED Step4-PASSED  - time date Thu Feb 11 11:30:07 2021-date Thu Feb 11 10:42:32 2021; exit: 0 0 0 0 0
1 1 1 1 1 tests passed, 0 0 0 0 0 failed
```

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

@tahuang1991

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

File: L1Trigger/CSCTriggerPrimitives/interface/CSCCathodeLCTProcessor.h [hunk 1]
```diff
@@ -203,6 +206,12 @@ class CSCCathodeLCTProcessor : public CSCBaseboard {
   std::vector<int> thePreTriggerBXs;
   std::vector<CSCCLCTPreTriggerDigi> thePreTriggerDigis;
 
+  /* data members for high multiplicity triggers */
+  void encodeHighMultiplicityBits(
+      const std::vector<int> halfstrip[CSCConstants::NUM_LAYERS][CSCConstants::NUM_HALF_STRIPS_7CFEBS]);
```
Comment (civanch) [inline L1Trigger/CSCTriggerPrimitives/interface/CSCCathodeLCTProcessor.h:211]: @dildick , it is possible to use "const std::vector<int>& ..." ?

File: L1Trigger/CSCTriggerPrimitives/interface/CSCAnodeLCTProcessor.h [hunk 2]
```diff
@@ -123,8 +123,9 @@ class CSCAnodeLCTProcessor : public CSCBaseboard {
   std::vector<CSCALCTPreTriggerDigi> thePreTriggerDigis;
 
   /* data members for high multiplicity triggers */
-  void encodeHighMultiplicityBits();
+  void encodeHighMultiplicityBits(const std::vector<int> wire[CSCConstants::NUM_LAYERS][CSCConstants::MAX_NUM_WIRES]);
```
Comment (civanch) [inline L1Trigger/CSCTriggerPrimitives/interface/CSCAnodeLCTProcessor.h:126]: @dildick , it seems, that reference to this vector would be better than vector itself.

Comment (dildick) [reply L1Trigger/CSCTriggerPrimitives/interface/CSCAnodeLCTProcessor.h:126]: @civanch correct. I will fix it.

File: L1Trigger/CSCTriggerPrimitives/interface/CSCCathodeLCTProcessor.h (hunk 1 above)
Comment (dildick) [reply L1Trigger/CSCTriggerPrimitives/interface/CSCCathodeLCTProcessor.h:211]: @civanch yes, I will fix it.

**Comment by dildick** [comment]
@civanch Implementing your suggestion does not seem to work. Otherwise there are arrays of references.

**Comment by dildick** [comment]
@rekovic Do you have any comments at this point?

**Comment by civanch** [comment]
@dildick , reference does work if an object are kept in the class. If an object is created in a class and transformed to another class reference cannot be used.

**Comment by dildick** [comment]
I'm sorry, I'm not following. The "wire" and "halfstrip" 2D arrays of vectors are passed between functions in the same class. That is how it was originally designed. They are not class members and are initialized once per event per CSC board. With that in mind, how would you suggest improving the code?

**Comment by dildick** [comment]
@civanch Do you have a suggestion?

**Comment by dildick** [comment]
@rekovic Do you have any comments?

**Comment by dildick** [comment]
Under discussion here https://github.com/cms-l1t-offline/cmssw/pull/891

**Pull request closed without merging**